v2g-liberty/rootfs/root/appdaemon/apps/v2g_liberty/modbus_evse_client.py

```diff
@@ -803,80 +838,131 @@ async def __get_and_process_registers(self, entities: list):
                     f"not type == int : ignored."
                 )
                 continue
-            if not entity["minimum_value"] <= new_state <= entity["maximum_value"]:
-                # self.__log(f"__get_and_process_registers: New value '{new_state}' for entity '{entity_name}' out of range {entity['minimum_value']} - {entity['maximum_value']} ignored.")
-                continue
 
-            await self.__update_entity(entity=entity, value=new_state)
-            await self.__update_state(entity_id=entity_name, state=new_state)
+            if not (entity["minimum_value"] <= new_state <= entity["maximum_value"]):
+                if entity["current_value"] is None:
+                    # This is very rare: current_value will only be None at startup.
+                    # Not setting a value will cause the application to hang, so even though
+                    # not correct still give it these values.
+                    if new_state < entity["minimum_value"]:
+                        best_guess = entity["minimum_value"]
+                    else:
+                        best_guess = entity["maximum_value"]
+                    self.__log(
+                        f"New value {new_state} for entity '{entity_name}' "
+                        f"out of range {entity['minimum_value']} "
+                        f"- {entity['maximum_value']} but current value is None, so"
+                        f"best guess for new value = {best_guess}."
+                    )
+                    new_state = best_guess
+                else:
+                    # self.__log(f"New value {new_state}' for entity '{entity_name}' "
+                    #            f"out of range {entity['minimum_value']} "
+                    #            f"- {entity['maximum_value']} ignored.")
+                    continue
+
+            await self.__update_ha_and_evse_entity(
+                evse_entity=entity, new_value=new_state
+            )
         return
 
-    async def __update_entity(self, entity: dict, value):
-        # self.__log(f"__update_entity called for {entity['ha_entity_name']} with value '{value}'.")
-        current_value = entity["current_value"]
-        if current_value != value:
-            entity["current_value"] = value
-            entity["previous_value"] = current_value
-            # call handler if defined
-            if "change_handler" in entity.keys():
-                str_action = entity["change_handler"]
+    async def __update_ha_and_evse_entity(self, evse_entity, new_value):
+        """Helper method to update both the evse- and ha-entity at the same time"""
+        await self.__update_evse_entity(evse_entity=evse_entity, new_value=new_value)
+        entity_id = f"sensor.{evse_entity['ha_entity_name']}"
+        await self.__update_ha_entity(entity_id=entity_id, new_value=new_value)
+
+    async def __update_evse_entity(self, evse_entity: dict, new_value):
```

Comment:
The name of this method confuses me, as it does not write a value to the EVSE, or am I missing something?

Reply:
Yes, it writes:
    evse_entity["current_value"] = new_value
    evse_entity["previous_value"] = current_value